src/game/loop.rs

```diff
@@ -0,0 +1,234 @@
+use crate::graphics::{Window, WindowSettings};
+use crate::input::{self, gamepad, keyboard, mouse, window, Input};
+use crate::load::{Join, LoadingScreen, Task};
+use crate::{Result, Timer};
+
+pub trait Loop<Game: super::Game> {
+    type Attributes;
+
+    fn new(
+        configuration: Self::Attributes,
+        game: &mut Game,
+        window: &Window,
+    ) -> Self;
+
+    fn load(window: &Window) -> Task<Self::Attributes>;
+
+    fn on_input(&mut self, input: &mut Game::Input, event: input::Event) {
+        input.update(event);
+    }
+
+    fn after_draw(
+        &mut self,
+        _game: &mut Game,
+        _input: &mut Game::Input,
+        _window: &mut Window,
+        _debug: &mut Debug,
+    ) {
+    }
+
+    fn run(window_settings: WindowSettings) -> Result<()>
+    where
+        Self: 'static + Sized,
+        Game: 'static,
+        Game::Input: 'static,
+    {
+        // Window creation
+        let event_loop = winit::event_loop::EventLoop::new();
+        let mut window = Window::new(window_settings, &event_loop)?;
+        let mut debug = Debug::new(window.gpu());
+
+        // Loading
+        debug.loading_started();
+        let (mut game, configuration) = {
+            let mut loading_screen = Game::LoadingScreen::new(window.gpu())?;
+
+            loading_screen.run(
+                (Game::load(&window), Self::load(&window)).join(),
+                &mut window,
+            )?
+        };
+
+        let mut game_loop = Self::new(configuration, &mut game, &mut window);
+        let mut input = Game::Input::new();
+        let mut gamepads = gamepad::Tracker::new();
+        debug.loading_finished();
+
+        let mut timer = Timer::new(Game::TICKS_PER_SECOND);
+
+        // Initialization
+        debug.frame_started();
+        timer.update();
+        window.request_redraw();
+
+        event_loop.run(move |event, _, control_flow| match event {
+            winit::event::Event::NewEvents(_) => {
+                debug.interact_started();
+            }
+            winit::event::Event::MainEventsCleared => {
+                if let Some(tracker) = &mut gamepads {
+                    while let Some((id, event, time)) = tracker.next_event() {
+                        game_loop.on_input(
+                            &mut input,
+                            input::Event::Gamepad { id, event, time },
+                        );
+                    }
+                }
+
+                game.interact(&mut input, &mut window);
+                input.clear();
+                debug.interact_finished();
+
+                if timer.tick() {
+                    debug.update_started();
+                    game.update(&window);
+                    debug.update_finished();
+                }
+
+                if game.is_finished() {
+                    *control_flow = winit::event_loop::ControlFlow::Exit;
+                }
+            }
+            winit::event::Event::RedrawRequested { .. } => {
```

Comment:
@hecrj As of b5db6813232f60f0bd040f62d289c81bf278de85, this doesn't seem to get called on Windows 10 except for the window launch and termination.

At the risk of asking a stupid question, I was wondering if the `request_redraw` call [here](https://github.com/hecrj/coffee/pull/62/files#diff-aba51a95c07772623d661e3692bd16d5R115-R117) should happen inside the `winit::event::Event::MainEventsCleared` instead. It looks like this was how it used to work and it doesn't look like it got changed during the `RedrawRequested` changes in `winit`. At least, the example seems to be left as is [here](https://github.com/rust-windowing/winit/commit/da4e6295a1c563400e62957e4d389340d8eda4e8#diff-b4aea3e418ccdb71239b96952d9cddb6L55).

Reply:
You are absolutely right!

I think I changed this for #63 because requesting redraws many times per frame caused erratic behavior on Wasm. IIRC, this should be fixed now in `winit`.

Thank you for the catch! :clinking_glasses: I pushed a commit with the fix.

Reply:
Thanks for the fix! I tried a couple examples in Windows 10 and everything seems to be working as expected 🤞